# Student Name: Wallace Esteves
# Student ID : 2020326


# Task 1

In [1]:
import pandas as pd

# Load the dataset
file_path = 'Data/Marge_Bedfordshire.csv'
data = pd.read_csv(file_path)

# Display the first few rows of the dataset
data.head()

Unnamed: 0                                           Crime ID    Month  \
0           0  f3b4eb5ff671847a8de4838f2deb31ad16825558ae1104...  2020-01   
1           1  ea45d7028058f7611f0ea7885c5003b253c4adf54b1df2...  2020-01   
2           2  4a9cf8bfd71238a0f5dfab36bc8d16e84afb7d1970879a...  2020-01   
3           3  4e28320826593de789a97acea21d0658a0bae95fc6ebbb...  2020-01   
4           4  3456382b4426f30a4d1d8aae3f10b6b7614dbb566946f8...  2020-01   

           Reported by         Falls within  Longitude   Latitude  \
0  Bedfordshire Police  Bedfordshire Police  -0.678572  51.960550   
1  Bedfordshire Police  Bedfordshire Police  -0.600189  51.873389   
2  Bedfordshire Police  Bedfordshire Police  -0.600189  51.873389   
3  Bedfordshire Police  Bedfordshire Police  -0.417743  52.237802   
4  Bedfordshire Police  Bedfordshire Police  -0.411488  52.225940   

                     Location  LSOA code            LSOA name  \
0  On or near Park/Open Space  E01017669  Aylesbury Vale 006A   
1    On or near Northall Road  E01017658  Aylesbury Vale 009D   
2    On or near Northall Road  E01017658  Aylesbury Vale 009D   
3   On or near Elm Tree Grove  E01017540         Bedford 001A   
4      On or near School Lane  E01017540         Bedford 001A   

                     Crime type  \
0     Criminal damage and arson   
1  Violence and sexual offences   
2  Violence and sexual offences   
3     Criminal damage and arson   
4                   Other theft   

                           Last outcome category  Context  
0                      Status update unavailable      NaN  
1                      Status update unavailable      NaN  
2                      Status update unavailable      NaN  
3  Investigation complete; no suspect identified      NaN  
4  Investigation complete; no suspect identified      NaN

# Task B: Identifying Variable Types and Missing Values

Explanation: We will categorize the variables into categorical, discrete, and continuous types based on their nature and the data they hold. We'll also check for missing values in each column.

In [2]:
# Identifying variable types and checking for missing values
variable_types = {
    "Categorical": [],
    "Discrete": [],
    "Continuous": []
}
missing_values = data.isnull().sum()

# Categorizing variables
for column in data.columns:
    if data[column].dtype == 'object':
        variable_types["Categorical"].append(column)
    elif data[column].dtype in ['int64', 'float64']:
        if data[column].nunique() < 10:
            variable_types["Discrete"].append(column)
        else:
            variable_types["Continuous"].append(column)

# Displaying the results
variable_types, missing_values


({'Categorical': ['Crime ID',
   'Month',
   'Reported by',
   'Falls within',
   'Location',
   'LSOA code',
   'LSOA name',
   'Crime type',
   'Last outcome category'],
  'Discrete': ['Context'],
  'Continuous': ['Unnamed: 0', 'Longitude', 'Latitude']},
 Unnamed: 0                    0
 Crime ID                  34876
 Month                         0
 Reported by                   0
 Falls within                  0
 Longitude                     0
 Latitude                      0
 Location                      0
 LSOA code                     0
 LSOA name                     0
 Crime type                    0
 Last outcome category     34876
 Context                  156767
 dtype: int64)

# Task C: Statistical Parameters Calculation
Explanation:
Now, we will calculate statistical parameters such as mean, median, minimum, maximum, and standard deviation for each of the numerical variables.

In [3]:
#Task C
# Calculating statistical parameters for continuous variables
statistical_params = data.describe().T[['mean', '50%', 'min', 'max', 'std']].rename(columns={'50%': 'median'})

# Displaying the results
statistical_params


mean        median        min            max           std
Unnamed: 0  78383.000000  78383.000000   0.000000  156766.000000  45254.879162
Longitude      -0.455201     -0.457553  -5.540364       1.731959      0.104470
Latitude       51.990858     51.917292  50.119077      55.234600      0.127505
Context              NaN           NaN        NaN            NaN           NaN

# Task D

In [4]:
#Task D
from sklearn.preprocessing import MinMaxScaler, StandardScaler, RobustScaler

# Selecting numerical columns for scaling
numerical_cols = ['Longitude', 'Latitude']

# Applying Min-Max Normalization
min_max_scaler = MinMaxScaler()
data_min_max = min_max_scaler.fit_transform(data[numerical_cols])

# Applying Z-score Standardization
standard_scaler = StandardScaler()
data_standardized = standard_scaler.fit_transform(data[numerical_cols])

# Applying Robust Scaler
robust_scaler = RobustScaler()
data_robust = robust_scaler.fit_transform(data[numerical_cols])

# Converting the scaled data back to a DataFrame for visualization
scaled_data = {
    "Min-Max Scaled": data_min_max,
    "Z-score Scaled": data_standardized,
    "Robust Scaled": data_robust
}
scaled_data_df = {key: pd.DataFrame(value, columns=numerical_cols) for key, value in scaled_data.items()}

# Displaying the first few rows of the scaled data
scaled_data_df["Min-Max Scaled"].head(),scaled_data_df["Z-score Scaled"].head(),scaled_data_df["Robust Scaled"].head()


(   Longitude  Latitude
 0   0.668534  0.359977
 1   0.679312  0.342939
 2   0.679312  0.342939
 3   0.704400  0.414176
 4   0.705260  0.411857,
    Longitude  Latitude
 0  -2.138151 -0.237702
 1  -1.387853 -0.921291
 2  -1.387853 -0.921291
 3   0.358556  1.936738
 4   0.418430  1.843706,
    Longitude  Latitude
 0  -2.846862  0.183725
 1  -1.837240 -0.186464
 2  -1.837240 -0.186464
 3   0.512778  1.361266
 4   0.593346  1.310886)

In [ ]:
# Here I decided to create a plot to have a better visualization

import matplotlib.pyplot as plt

# Function to create scatter plots for scaled data
def create_scatter_plot(dataframe, title, ax):
    ax.scatter(dataframe['Longitude'], dataframe['Latitude'])
    ax.set_title(title)
    ax.set_xlabel('Scaled Longitude')
    ax.set_ylabel('Scaled Latitude')

# Creating subplots
fig, axes = plt.subplots(1, 3, figsize=(18, 6))

# Min-Max Scaled plot
create_scatter_plot(scaled_data_df["Min-Max Scaled"], 'Min-Max Scaled Data', axes[0])

# Z-score Scaled plot
create_scatter_plot(scaled_data_df["Z-score Scaled"], 'Z-score Scaled Data', axes[1])

# Robust Scaled plot
create_scatter_plot(scaled_data_df["Robust Scaled"], 'Robust Scaled Data', axes[2])

# Displaying the plots
plt.tight_layout()
plt.show()


This code will apply the three scaling techniques to the 'Longitude' and 'Latitude' columns and then convert the scaled data back into pandas DataFrames for ease of visualization. Let's execute it.

Results from Task D: Normalization and Standardization
The numerical data (Longitude and Latitude) has been scaled using three different techniques:

Min-Max Scaled Data:

Scaled such that the values fall within the range [0, 1].
Example values:
Longitude: [0.6685, 0.6793, 0.6793, 0.7044, 0.7053]
Latitude: [0.3600, 0.3429, 0.3429, 0.4142, 0.4119]
Z-score Scaled Data:

Standardized with a mean of 0 and a standard deviation of 1.
Example values:
Longitude: [-2.1382, -1.3879, -1.3879, 0.3586, 0.4184]
Latitude: [-0.2377, -0.9213, -0.9213, 1.9367, 1.8437]
Robust Scaled Data:

Scaled using the median and interquartile range, making it more robust to outliers.
Example values:
Longitude: [-2.8469, -1.8372, -1.8372, 0.5128, 0.5933]
Latitude: [0.1837, -0.1865, -0.1865, 1.3613, 1.3109]
These transformations are crucial for data analysis and modeling, as they standardize the range and distribution of the data, making it easier to compare and understand.